array/array.go

```diff
@@ -0,0 +1,144 @@
+package array
+
+import (
+	"errors"
+)
+
+const (
+	// EmptyValue is a constant indicates that
+	// a value returned from the Array is empty.
+	EmptyValue       = 0
+	defaultCapacity  = 8
+	growShrinkFactor = 2
+	upperLoadFactor  = 1.0
+	lowerLoadFactor  = 0.25
+	indexOutOfRange  = "Index is out of range"
+)
+
+// Array is a struct for storing linear equal-size data.
+type Array struct {
+	items []int
+	size  int
+}
+
+// NewArray creates a new Array instance.
+func NewArray(args ...int) *Array {
+	capacity := defaultCapacity
+	if args != nil && args[0] > 0 {
+		capacity = args[0]
+	}
+	return &Array{
+		items: make([]int, capacity),
+	}
+}
+
+// Size returns the number of items being stored in the Array.
+func (a *Array) Size() int {
+	return a.size
+}
+
+// Capacity returns the current maximum number of items
+// that can be stored in the Array.
+func (a *Array) Capacity() int {
+	return len(a.items)
+}
+
+// IsEmpty returns whether the Array has no items or not.
+func (a *Array) IsEmpty() bool {
+	return a.size == 0
+}
+
+// ValueAt returns the value stored at the given index.
+func (a *Array) ValueAt(index int) (int, error) {
+	if index < 0 || index >= a.size {
+		return EmptyValue, errors.New(indexOutOfRange)
+	}
+	return a.items[index], nil
+}
+
+// Insert adds the value to the given index and shift
+// the items from that index to the right.
+func (a *Array) Insert(value int, index int) error {
+	if index < 0 || index >= a.size {
```

Comment:
It should be `index < 0 || index > a.size`. It's ok to insert an item at index 8 while array is having 8 items.